File: java/src/main/java/ch/guengel/Array.java
package ch.guengel;

public class Array {
    private static final int NUMBER_OF_ELEMENTS = 1_000_000;
    private Struct[] structArray = new Struct[NUMBER_OF_ELEMENTS];
    private long[] sumArray = new long[NUMBER_OF_ELEMENTS];

    public void fillArray() {
        for (var i = 0; i < NUMBER_OF_ELEMENTS; i++) {
            structArray[i] = new Struct(i, i + 1, i + 2);
        }
    }

    public void fillSumArray() {
        Struct struct;
        for (var i = 0; i < NUMBER_OF_ELEMENTS; i++) {
            struct = structArray[i];
            sumArray[i] = struct.valueA +
                    struct.valueB +
                    struct.valueC;
        }
    }

    long sumOfArray() {
        var sum = 0L;
        for (var subSum: sumArray) {
            sum += subSum;
        }
        return sum;
    }

}

File: cpp/array.cc
#include <memory>

#include "array.hh"

void Array::fillStructArray() {
    for (auto i = 0; i < NUMBER_OF_ELEMENTS; i++) {
        StructA& ref = structArray[i];
        ref.valueA = i;
        ref.valueB = i + 1;
        ref.valueC = i + 2;
    }
}

void Array::fillSumArray() {
    StructA* ptr;
    for (auto i = 0; i < NUMBER_OF_ELEMENTS; i++) {
        ptr = structArray + i;
        sumArray[i] = ptr->valueA + ptr->valueB + ptr->valueC;
    }
}

long Array::sumOfArray() const {
    auto sum = 0l;
    for (auto i = 0; i < NUMBER_OF_ELEMENTS; i++) {
        sum += sumArray[i];
    }
    return sum;
}
